content/blog/2021/05/multicast-failure-routing/index.md

```diff
@@ -0,0 +1,199 @@
+            newExchange.getIn().setBody(aggregatedDataList);
+            return newExchange;
+        }
+
+        List<String> oldData = oldExchange.getIn().getBody(List.class);
+        oldData.add(newExchange.getIn().getBody(String.class));
+        oldExchange.getIn().setBody(oldData);
+
+        return oldExchange;
+    }
+}
+```
+
+On executing the same, we see following logs
+```log
+2021-05-06 12:43:18.565  INFO 13956 --- [qtp916897446-42] route1                                   : received request
+2021-05-06 12:43:18.566  INFO 13956 --- [qtp916897446-42] route1                                   : Entering multicast
+2021-05-06 12:43:18.575  INFO 13956 --- [ #4 - Multicast] route2                                   : Executing PATH_1 - exception path
+2021-05-06 12:43:18.575  INFO 13956 --- [ #4 - Multicast] route2                                   : Starting exception throw
+2021-05-06 12:43:18.578  INFO 13956 --- [ #4 - Multicast] route2                                   : Exception handler invoked
+2021-05-06 12:43:18.579  INFO 13956 --- [ #4 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator {"data" : "err"}
+2021-05-06 12:43:19.575  INFO 13956 --- [ #3 - Multicast] route3                                   : Executing PATH_2 - success path
+2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] route3                                   : PATH_2
+2021-05-06 12:43:21.576  INFO 13956 --- [ #3 - Multicast] c.e.d.m.SimpleFlowMergeAggregator        : Inside aggregator DATA_FROM_PATH_2
+```
+
+## What could take you by a surprise?
+
+- When the multicast completes aggregating exchanges from child branches, one might intermittently note that it stops routing the remaining processors (those 2 additional log and a transform step in above example). On execution tracing exercise, you will notice this happens in a special scenario when the very first exchange which arrives in the aggregator (from first completed child branch) had encountered an exception during its course or/and was handled via onException flows. On the flip side, if the first exchange was a successful and even though all the remaining ones experienced a failure, it still continued routing the remaining processors/steps.
+
+
+## Analysis
+
+To understand this better, lets deep dive into the open source codebase. Notice the [Pipeline class]( https://github.com/apache/camel/blob/6dff85675e9b73e8a528bc2683935ec3c1ed26b7/core/camel-core-processor/src/main/java/org/apache/camel/processor/Pipeline.java ). PipelineProcessor (part of camel core framework processors) performs an evaluation after every user processors (user added steps in a camel flow) on whether it should continue routing to the next processor.
+This decision is made inside Pipeline class on [this if block](https://github.com/apache/camel/blob/6dff85675e9b73e8a528bc2683935ec3c1ed26b7/core/camel-core-processor/src/main/java/org/apache/camel/processor/Pipeline.java#L79) <br>
```

Comment:
Yeah I see the issue, the `main` branch could diverge and this line would not be valid any more. Do you think it makes sense to provide a snippet of the code from `Pipeline.java` here instead?

Reply:
Yes makes sense

Reply:
done. can you review now?